libpromises/cmdb.c

```diff
@@ -0,0 +1,246 @@
+            EvalContextVariablePut(ctx, &ref, data, CF_DATA_TYPE_CONTAINER, "source=cmdb");
+        }
+    }
+    return true;
+}
+
+static bool ReadCMDBClasses(EvalContext *ctx, JsonElement *classes)
+{
+    assert(classes != NULL);
+
+    if (JsonGetType(classes) != JSON_TYPE_OBJECT)
+    {
+        Log(LOG_LEVEL_ERR, "Invalid 'classes' CMDB data, must be a JSON object");
+        return false;
+    }
+
+    if (!JsonWalk(classes, CheckObjectForUnexpandedVars, NULL, CheckPrimitiveForUnexpandedVars, NULL))
+    {
+        Log(LOG_LEVEL_ERR, "Invalid 'classes' CMDB data, cannot contain variable references");
+        return false;
+    }
+
+    JsonIterator iter = JsonIteratorInit(classes);
+    while (JsonIteratorHasMore(&iter))
+    {
+        const char *key = JsonIteratorNextKey(&iter);
+        JsonElement *data = JsonObjectGet(classes, key);
+        if (JsonGetElementType(data) == JSON_ELEMENT_TYPE_PRIMITIVE)
+        {
+            const char *expr = JsonPrimitiveGetAsString(data);
+            if (!StringEqual(expr, "any::"))
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Invalid class specification '%s' in CMDB data, only \"any::\" allowed", expr);
+                continue;
+            }
+            Log(LOG_LEVEL_VERBOSE, "Installing CMDB class '%s'", key);
+            EvalContextClassPutSoftNS(ctx, "cmdb", key, CONTEXT_SCOPE_NAMESPACE, "source=cmdb");
+        }
+        else if (JsonGetContainerType(data) == JSON_CONTAINER_TYPE_ARRAY &&
+                 JsonArrayContainsOnlyPrimitives(data))
+        {
+            if ((JsonLength(data) != 1) ||
+                (!StringEqual(JsonPrimitiveGetAsString(JsonArrayGet(data, 0)), "any::")))
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Invalid class specification '%s' in CMDB data, only '[\"any::\"]' allowed",
+                    JsonPrimitiveGetAsString(JsonArrayGet(data, 0)));
+                continue;
+            }
+            Log(LOG_LEVEL_VERBOSE, "Installing CMDB class '%s'", key);
+            EvalContextClassPutSoftNS(ctx, "cmdb", key, CONTEXT_SCOPE_NAMESPACE, "source=cmdb");
+        }
+        else
+        {
+            Log(LOG_LEVEL_ERR, "Invalid CMDB class data for class '%s', must be a JSON object", key);
+        }
+    }
+    return true;
+}
+
+bool LoadCMDBData(EvalContext *ctx)
+{
+    char file_path[PATH_MAX] = {0};
+    strncpy(file_path, GetDataDir(), sizeof(file_path));
+    JoinPaths(file_path, sizeof(file_path), HOST_SPECIFIC_DATA_FILE);
+    if (access(file_path, F_OK) != 0)
+    {
+        Log(LOG_LEVEL_VERBOSE, "No host-specific JSON data available at '%s'", file_path);
+        return true;            /* not an error */
+    }
+    if (access(file_path, R_OK) != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Cannot read host-spefic JSON data from '%s'",
+            file_path);
+        return false;
+    }
+
+    JsonElement *data = ReadJsonFile(file_path, LOG_LEVEL_ERR, HOST_SPECIFIC_DATA_MAX_SIZE);
```

Comment:
I can only see json element `data` destroyed in case of error, on line 216. Should it not be destroyed before we return from this function no matter what?

Reply:
Good catch!